### PR TITLE
subgraph

### PR DESCRIPTION
**Pull request by CarloLucibello**
... since it can be handy to map back the vertices to the original graph

also made some internal change to induce_subgraph, in particular I eliminated the dummy return when `length(iter) == nv(g)` since this is  a case when one wants to create a graph isomorphism

### CODE REVIEW

**Comment by CarloLucibello** [comment]
this is eventually good to merge

**Comment by sbromberger** [comment]
I don't like the name. It's not an induced subgraph, really, since you're just removing the edges of the non-selected vertices, and it's fundamentally different from `induced_subgraph` (which also has a helper function `[]`). Could we call it `filter_edges!` or something else? I don't necessarily want to bikeshed the name but it needs to be different.

File: src/operators.jl [hunk 1]
```diff
@@ -184,20 +184,39 @@ Replicate `len` times `h` and connect each vertex with its copies in a path
...
+end
+
 """
     induced_subgraph(g, iter)
 
-Filters graph `g` to include only the vertices present in the iterable
-argument `vs`. Returns the subgraph of `g` induced by `vs`.
+Returns the subgraph of `g` induce by the vertices in the iterable
+argument `iter`.
+
```
Comment (sbromberger) [inline src/operators.jl]: Please leave the first sentence. It's clearer.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# include("../src/LightGraphs.jl")
+include("../src/LightGraphs.jl")
```
Comment (sbromberger) [inline test/runtests.jl]: we shouldn't need this.

**Comment by IssamT** [comment]
How about something like this ? (the name I put is just an example but you can replace it with anything other than induced_subgraph).

```
function induced_subgraph_and_vmap(g::SimpleGraph, iter)
    vmap = Vector{Int}(length(iter))
    resize!(vmap, length(iter))
    for (i, v) in enumerate(iter)
        vmap[i] = v
    end
    return induced_subgraph(g, iter), vmap
end
```

I suggest this because of 2 reasons:

1 - I know @CarloLucibello `induced_subgraph!` is defined with the exclamation mark because it modifies one of the attributes namely `vmap`. but a first understanding from a user who reads `induced_subgraph!` and `induced_subgraph` could be that one modifies the graph itself while the other returns a new graph object (for the induced subgraph).

2 - Is a user really required to create the `vmap` and pass it to the function? If we add it just to have a new signature which is needed for overloading the function then in my opinion it is better to just change the name.

**Comment by sbromberger** [comment]
This is essentially equivalent to getting the in- and out_edges of a set of vertices and then calling `rem_edge!()` on each. It's not inducing a subgraph. It's a convenience function, but my strong opinion is that it should not be confused for any existing method, which is what any name starting with `induced` will do.

**Comment by CarloLucibello** [comment]
@sbromberger from your last two comments I think you are getting this wrong: we are not filtering edges here, the behaviour is exactly the same as the old induce_supgraph. The returned graph has length(iter) nodes. This is only about adding a convenience function wich gives also a map between the new and the old nodes, as it was requested for instance in #463 

@IssamT I agree with your arguments. I don't like `induced_subgraph_and_vmap` too much but also I can't think of something better, so I'll go with that

File: src/operators.jl [hunk 3]
```diff
@@ -184,20 +184,39 @@ Replicate `len` times `h` and connect each vertex with its copies in a path
...
 
+
+"""
+    induced_subgraph!(vmap::Vector{Int}, g, iter)
+
+Returns [`induced_subgraph`](@ref)`(g, iter)` and fills
+`vmap` so that the `i` vertex in the subgraph corresponds to
+the vertex `vmap[i]` in `g`.
+"""
+function induced_subgraph!(vmap::Vector{Int}, g::SimpleGraph, iter)
```
Comment (sbromberger) [inline src/operators.jl]: Needs a type on `iter`

**Comment by sbromberger** [comment]
I got it. However, see line notes for some comments.

Given this, I'd suggest calling this `induced_subgraph_with_map`.

**Comment by sbromberger** [comment]
Not to derail, but perhaps we should roll this into `induced_subgraph`. I don't want to make it type unstable, but maybe there's a way to do this.

**Comment by CarloLucibello** [comment]
I tried to incorporate all the suggestions and also simplified the name in `subgraph`

Note that since all the real work is done in _subgraph, return type instability doesn't affect performance

**Comment by CarloLucibello** [comment]
ok, I found a more satifying solution
- subgraph returns sg, vmap
- g[iter]  returns only  the subgraph

also I added the possibility to create a subgraph from a vectors of edges

**Comment by sbromberger** [comment]
I don't like changing the name to `subgraph`. "Induced subgraph" has a specific definition and one that we use. Can we keep `induced_subgraph`, but have it return a vmap, and then use `[]` as you have it just returning the (di)graph?

We won't need `deprecated.jl` or `subgraph.jl` - this should remain in `operators.jl`.

**Comment by sbromberger** [comment]
Why did you edit my comment?

File: src/deprecated.jl [hunk 4]
```diff
@@ -0,0 +1,3 @@
+@deprecate rev(e::Edge) reverse(e)
```
Comment (sbromberger) [inline src/deprecated.jl]: We can get rid of this now - it's been there for at least one major version.

File: src/deprecated.jl [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+@deprecate induced_subgraph subgraph
```
Comment (sbromberger) [inline src/deprecated.jl]: I don't agree with deprecating `induced_subgraph`. We should keep this name.

**Comment by CarloLucibello** [comment]
>Why did you edit my comment?

Ops, I clicked the wrong buttons in answeing you comment. Did you read my reply? it's gone now

**Comment by sbromberger** [comment]
I did read it. I'm not convinced - induced subgraph means something specific, and we offer it.

**Comment by CarloLucibello** [comment]
and what about deprecated.jl and subgraphs.jl? I also commented on that...

**Comment by sbromberger** [comment]
I left line comments in deprecated. Basically it should be a separate PR. I'm not convinced we need it for a single remaining function.

For subgraphs.jl, I'd prefer keeping things in operators.jl unless we plan on having more ways to get subgraphs.

**Comment by CarloLucibello** [comment]
ok, incorporated the comments, ready to merge

File: src/operators.jl [hunk 6]
```diff
@@ -309,3 +274,104 @@ function tensor_product{G<:SimpleGraph}(g::G, h::G)
...
+
+sg, vmap = subgraph(g, [2,8,3,4])
+@asssert sg == g[[2,8,3,4]]
+
+elist = [Edge(1,2), Edge(3,4), Edge(4,8)]
+sg, vmap = subgraph(g, elist)
+@asssert sg == g[elist]
+```
+"""
+function induced_subgraph{T<:SimpleGraph}(g::T, vlist)
```
Comment (sbromberger) [inline src/operators.jl]: vlist needs type annotation. `Vector{Int}` ?

File: test/runtests.jl [hunk 7]
```diff
@@ -2,13 +2,10 @@
...
 using Base.Test
 
-
 g1 = smallgraph(:petersen)
 g2 = smallgraph(:tutte)
 g3 = PathGraph(5)
 g4 = PathDiGraph(5)
-g5 = DiGraph(4)
-add_edge!(g5,1,2); add_edge!(g5,2,3); add_edge!(g5,1,3); add_edge!(g5,3,4)
 g6 = smallgraph(:house)
```
Comment (sbromberger) [inline test/runtests.jl]: The reason these are in runtests is so that all subtests can reuse them. Why are we pulling g5 out, and if we do, shouldn't we renumber g6?

File: test/traversals/bfs.jl [hunk 8]
```diff
@@ -1,9 +1,12 @@
+g5 = DiGraph(4)
+add_edge!(g5,1,2); add_edge!(g5,2,3); add_edge!(g5,1,3); add_edge!(g5,3,4)
+
 z = bfs_tree(g5, 1)
 visitor = LightGraphs.TreeBFSVisitorVector(zeros(Int,nv(g5)))
 LightGraphs.bfs_tree!(visitor, g5, 1)
 t = visitor.tree
-@test nv(z) == 4 && ne(z) == 3 && !has_edge(z, 2, 3)
 @test t == [1,1,1,3]
+@test nv(z) == 4 && ne(z) == 3 && !has_edge(z, 2, 3)
```
Comment (sbromberger) [inline test/traversals/bfs.jl]: what happened here?

File: src/operators.jl (hunk 6 above)
Comment (CarloLucibello) [inline src/operators.jl]: we need also ranges, I would live this untyped as it was before

File: test/runtests.jl (hunk 7 above)
Comment (CarloLucibello) [inline test/runtests.jl]: this is bad practice, object definition should stay near where they are used. I run into errors when defined a g5 graph to add a test in subgraph, and this lead to failures in successively included files. It took a while to figure out the cause of that

File: src/operators.jl (hunk 6 above)
Comment (sbromberger) [inline src/operators.jl]: The common supertype is now `AbstractArray{Int64,1}` - this was originally omitted since there was no common supertype. I realize we're making a new change, but let's add the annotation.

File: test/traversals/bfs.jl (hunk 8 above)
Comment (CarloLucibello) [inline test/traversals/bfs.jl]: nothing, just moved one line down in order to do some debugging (see the previous comment)

File: test/runtests.jl (hunk 7 above)
Comment (sbromberger) [inline test/runtests.jl]: I'd argue that it's not really "bad practice", but that will derail the conversation. Instead, let's not address it piecemeal - my suggestion is to leave it in runtests and then, if it offends you, make a PR that fixes all the cases.

File: test/traversals/bfs.jl (hunk 8 above)
Comment (sbromberger) [inline test/traversals/bfs.jl]: ok. It seems really unrelated to the PR though, which is what piqued my curiosity.

File: test/runtests.jl (hunk 7 above)
Comment (CarloLucibello) [inline test/runtests.jl]: I don't  like like to waste time in reverting a step in the right direction (double the the waste) anyway, will do

**Comment by CarloLucibello** [comment]
hopefully that should be it

**Comment by sbromberger** [comment]
Yup, looks great. Thanks.
